File: 0062-unique-paths/0062-unique-paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<int> next(n,0);
    
        
        for(int i=m-1;i>=0;i--){
            vector<int> curr(n,0);
            for(int j=n-1;j>=0;j--){
                if(i==m-1 && j==n-1) curr[j]=1;
                else{
                    curr[j]=(i+1<m?next[j]:0)+(j+1<n?curr[j+1]:0);
                }
            }
            next=curr;
        }
        return next[0];

    }
};

File: 0062-unique-paths/0062-unique-paths.py
class Solution:
    def __init__(self): 
        self.res=0
    def dfs(self,dp,m,n,i,j):
        if i==m or j==n:
            return 0
        if i==m-1 and j==n-1:
            return 1
        if dp[i][j]!=-1: 
            return dp[i][j]
        dp[i][j]=self.dfs(dp,m,n,i+1,j)+self.dfs(dp,m,n,i,j+1)
        return dp[i][j]

    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[-1]*n for _ in range(m)]
        return self.dfs(dp,m,n,0,0)
        # return self.res
